### PR TITLE
Time dependent flag support (with some extras)

### PR DESCRIPTION
**Pull request by aewallwi**
The primary goal of this PR is to add support for time-dependent flags to `hera_pspec`. Implementing this in a way that as much a possible keeps runtimes reasonable required adding significant caching machinery and touches (but does not substantially reorganize) the core code. This warrants a detailed conversation amongst stakeholders and probably some more detailed tests on the side (on top of the standard unittests).

Caching includes `R`, `M`, `W`, `G`, `H`, `E`, and `unnormed_V` matrices. I hash these matrices with some new hashing functions; `weighting_hash`, `fisher_hash` and `cov_hash` in the lines

https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L504 
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L572
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L605

these functions hash matrices against appropriate sets of analysis settings, flagging patterns (so as to necessarily repeat per-time calculation) and baseline (pairs) where appropriate. 

`Y` has been significantly changed
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1079
`R` has been signifcantly changed 
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1156
`get_H`
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1855
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1855
and a new helper function `_get_H`

`get_G` and helper `_get_G`
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1698
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1618
 
`get_MW` has been eliminated in exchange for two separate functions and helpers
`get_M` / `_get_M`
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L2545
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L2377

`get_W` and `_get_W`
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L2496
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L2605

`get_unnormed_V` and `get_unnormed_E` also have caching added.
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1951
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L2240

The caching structure also involved some significant rewrite of the order of `pspecdata.pspec()`
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L3219
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1813

In order to facilitate speed up of each matrix calculation, I also significantly expanded where FFTs are used. They used to only be used in `q_hat` but I have added them to computing `H`, `G`, and `V`.  See lines
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1665

The use of `allow_fft` also used to be restricted to when `exact_norm` was `False`. I have adjusted `exact_norm` computations to allow for its use with `allow_fft`, see 
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecdata.py#L1612
 

A bit tangential. I also added an airy beam
https://github.com/HERA-Team/hera_pspec/blob/8c67c367bcac84c09fccea48b05f69e58925e820/hera_pspec/pspecbeam.py#L388

### CODE REVIEW

**Comment by anguta** [comment]
This PR addressed issues where `identity_weights` `G` and `H` matrices were cached  per baseline-pair which slowed things down a lot. @nkern raise an issue with `np.nan` values in window functions. I'm still parsing this issue and would like to engage start a back and forth to identify and remove the problem.

File: hera_pspec/pspecdata.py [hunk 1]
```diff
@@ -12,14 +12,18 @@
 import warnings
 import json
 import uvtools.dspec as dspec
+# This is a list of weights that will depend on baseline.
```
Comment (nkern) [inline hera_pspec/pspecdata.py:15]: give two line breaks after import statements

File: hera_pspec/pspecdata.py [hunk 2]
```diff
@@ -56,17 +60,24 @@ def __init__(self, dsets=[], wgts=None, dsets_std=None, labels=None,
 
         cal_flag : bool, optional
             If True, propagate flags from calibration into data
+
+        r_cache : dict, optional
```
Comment (nkern) [inline hera_pspec/pspecdata.py:64]: r_cache isn't actually a parameter in the class call

File: hera_pspec/pspecdata.py (hunk 1 above)
Comment (anguta) [reply hera_pspec/pspecdata.py:15]: I've moved this.

File: hera_pspec/pspecdata.py (hunk 2 above)
Comment (anguta) [reply hera_pspec/pspecdata.py:64]: removed this.

**Comment by anguta** [comment]
With separable flags, this code currently runs 50% slower then the master branch.

There does not appear to be a major culprit. 

image

image

**Pull request closed without merging**